consensus/ibft/src/main/java/org/hyperledger/besu/consensus/ibft/validation/MessageValidatorFactory.java

```diff
@@ -83,6 +95,18 @@ public MessageValidator createMessageValidator(
     final BlockValidator blockValidator =
         protocolSchedule.getByBlockNumber(roundIdentifier.getSequenceNumber()).getBlockValidator();
     final Collection<Address> validators = getValidatorsAfterBlock(parentHeader);
+    final Collection<Address> previousValidators = getValidatorsForBlock(parentHeader);
+
+    // Log any change in validators
+    if (!(validators.containsAll(previousValidators))
```

Comment:
Think this could end up logging multiple times for a height if had we multiple round changes. A better place is probably when we import the block in the IbftRound importBlockToChain method then we know the block is finalised

Reply:
Yeah I was wondering if this could cause more than 1 log entry, so I'll look again at where the check could be made to ensure only 1 log entry per block. The specific case we were trying to diagnose was the case where the validators had changed, but immediately afterwards too many of the validators went offline to mine a new block - so putting the log entry on `imortBlockToChain` wouldn't help in that scenario. I'll have a look at where the log entry could go that is before the block is produced, but doesn't give duplicate log entries.

Reply:
@jframe I've refactored the PR so that the validator change is checked and/or logged in `handleBlockTimerExpiry`, which seems a more sensible place to put a per-block check, and only writes a log entry for round `0`. That prevents multiple log entries if it takes several rounds to agree on a block.

I exposed the existing `MessageValiatorFactory.getValidators<For|After>Block()` methods as statics to avoid re-implementing them in `QbftBlockHeightManager/IbftBlockHeightManager`. Might be a cleaner way of re-using them - let me know what you think.